File: scripts/pipeline/assemble/senator-builder.mjs
import { log } from "../config.mjs";
import { validateSenator } from "./validator.mjs";

/**
 * Assemble a complete Senator record from all pipeline data.
 *
 * @param {Object} baseSenator - From normalize-members.mjs
 * @param {Object} funding - From normalize-finance.mjs
 * @param {Object} votingRecord - From normalize-votes.mjs (with cross-reference data)
 * @param {Array} lobbyingMatches - From cross-reference.mjs
 * @param {Object} corruptionScore - From score-calculator.mjs
 * @param {string} punkNickname - From nickname-generator.mjs
 * @returns {Object} Complete, validated Senator record
 */
export function buildSenator(
  baseSenator,
  funding,
  votingRecord,
  lobbyingMatches,
  corruptionScore,
  punkNickname
) {
  const senator = {
    ...baseSenator,
    punkNickname: punkNickname || baseSenator.punkNickname || "TBD",
    corruptionScore: corruptionScore || baseSenator.corruptionScore,
    funding: funding || baseSenator.funding,
    votingRecord: votingRecord || baseSenator.votingRecord,
    lobbyingMatches: lobbyingMatches || baseSenator.lobbyingMatches,
  };

  return validateSenator(senator);
}

File: backend/app/pipeline/assemble/senator_builder.py
"""
Senator builder — assembles a complete Senator record from all pipeline data.
"""

import logging

from app.pipeline.assemble.validator import validate_senator

logger = logging.getLogger(__name__)


def build_senator(
    base_senator: dict,
    funding: dict | None,
    voting_record: dict | None,
    lobbying_matches: list[dict] | None,
    corruption_score: dict | None,
) -> dict:
    """
    Assemble a complete Senator record from all pipeline data.

    Args:
        base_senator: From normalize-members (base senator record).
        funding: From normalize-finance.
        voting_record: From normalize-votes (with cross-reference data).
        lobbying_matches: From cross-reference analysis.
        corruption_score: From score-calculator.

    Returns:
        Complete, validated Senator record.
    """
    senator = {
        **base_senator,
        "representationScore": (
            corruption_score or base_senator.get("representationScore")
        ),
        "funding": funding or base_senator.get("funding"),
        "votingRecord": voting_record or base_senator.get("votingRecord"),
        "lobbyingMatches": (
            lobbying_matches or base_senator.get("lobbyingMatches")
        ),
    }

    return validate_senator(senator)
